Account/FxADeviceRegistrator.swift

```diff
@@ -0,0 +1,125 @@
+
+        public var description: String {
+            switch self {
+            case DeviceRegistrationFailed: return "Device registration failed."
+            }
+        }
+    }
+
+    public static func registerOrUpdateDevice(account: FirefoxAccount, state: MarriedState) -> Deferred<Maybe<FxADeviceRegistrationResult>> {
+        if let _ = account.fxaDeviceId where account.deviceRegistrationVersion == DeviceRegistrationVersion {
+            return Deferred(value: Maybe(success: FxADeviceRegistrationResult.AlreadyRegistered))
+        }
+
+        let client = FxAClient10(endpoint: account.configuration.authEndpointURL)
+        let sessionToken = state.sessionToken
+        let name = DeviceInfo.defaultClientName()
+        let device: FxADevice
+        if let deviceId = account.fxaDeviceId {
+            device = FxADevice.forUpdate(name, id: deviceId)
+        } else {
+            device = FxADevice.forRegister(name, type: "mobile")
+        }
+
+        return client.registerOrUpdateDevice(sessionToken, device: device).bind { result in
+            let deferred = Deferred<Maybe<FxADeviceRegistrationResult>>()
+
+            if let device = result.successValue,
+                let deviceId = device.id {
+                account.fxaDeviceId = deviceId
+                account.deviceRegistrationVersion = DeviceRegistrationVersion
+                deferred.fill(Maybe(success: FxADeviceRegistrationResult.RegisteredOrUpdated))
+                return deferred
+            }
+
+            let error = result.failureValue as? FxAClientError
+            switch (error) {
+            case let .Remote(remoteError)?:
+                switch (remoteError.code) {
+                case FxAccountRemoteError.DEVICE_SESSION_CONFLICT:
+                    recoverFromDeviceSessionConflict(account, client: client, deferred: deferred, sessionToken: sessionToken)
+                case FxAccountRemoteError.UNKNOWN_DEVICE:
+                    // TODO: Shouldn't this also clear registration version? Android doesn't.
+                    handleUnknownDevice(account)
+                case FxAccountRemoteError.INVALID_AUTHENTICATION_TOKEN:
+                    logErrorAndResetDeviceRegistrationVersion(account, description: String(remoteError))
+                    handleTokenError(account, client: client)
+                default:
+                    logErrorAndResetDeviceRegistrationVersion(account, description: String(remoteError))
+                }
+            case let .Local(localError)?:
+                self.logErrorAndResetDeviceRegistrationVersion(account, description: localError.description)
+            default:
+                // This shouldn't happen.
+                self.logErrorAndResetDeviceRegistrationVersion(account, description: "invalid error type")
+                assertionFailure()
+            }
+
+            deferred.fillIfUnfilled(Maybe(failure: DeviceRegistratorError.DeviceRegistrationFailed))
```

Comment:
Hm, we already handle and log the specific errors above in the `switch` statement. Is there anything more specific we'd want to do with these errors outside of this abstraction layer?